File: gpt-4-turbo/stage2/declare_directive_link_clause.c
#include <stdio.h>
#include <stdlib.h>
#include <openacc.h>

#ifndef T1
// T1:declare directive link clause,V:2.7-3.3

// Global variable to test link clause with
int global_test_var = 10;

// Declare link clause for the global variable
#pragma acc declare link(global_test_var)

int test1(){
    int err = 0;
    srand(42);

    // Original value of the global variable
    int original_value = global_test_var;
    int update_value = rand() % 100; // Random value to add

    // Modify global variable in a parallel region on the device
    #pragma acc parallel
    {
        #pragma acc loop
        for(int i = 0; i < 1; i++){
            global_test_var += update_value;
        }
    }

    // Update host variable by copying data from the device
    #pragma acc update self(global_test_var)

    if (global_test_var != original_value + update_value) {
        printf("Test failed: Expected %d, got %d\n", original_value + update_value, global_test_var);
        err++;
    } else {
        printf("Test passed: Global variable updated correctly.\n");
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
#ifndef T1
    if (test1() != 0){
        failcode = 1;
    }
#endif
    if (failcode == 0) {
        printf("All tests passed.\n");
    } else {
        printf("Some tests failed.\n");
    }
    return failcode;
}

File: gpt-4-turbo/stage2/declare_directive_link_clause.f90
MODULE mod_globals
  INTEGER :: global_var = 0
END MODULE mod_globals

! This declaration is needed for OpenACC
USE OPENACC
! Include the global variable module
USE mod_globals
IMPLICIT NONE

! Test function to verify declare directive's link clause
LOGICAL FUNCTION test_declare_link_clause()
  INTEGER :: local_var
  INTEGER :: errors = 0
  
  ! Set the device data for global_var using the declare directive with link clause
  !$ACC DECLARE LINK(global_var)
  
  ! Create a data region on the device for global_var and modify it
  !$ACC ENTER DATA COPYIN(global_var)
  !$ACC PARALLEL LOOP
  DO local_var = 1, 1
    global_var = 5
  END DO
  !$ACC EXIT DATA COPYOUT(global_var)
  
  ! Check if the modification is reflected on the host
  IF (global_var /= 5) THEN
    errors = errors + 1
  END IF
  
  ! Test passes if there are no errors
  IF (errors == 0) THEN
    test_declare_link_clause = .TRUE.
  ELSE
    test_declare_link_clause = .FALSE.
  END IF
END FUNCTION test_declare_link_clause

PROGRAM main
  USE OPENACC
  USE mod_globals
  IMPLICIT NONE
  LOGICAL :: test_result
  
  ! Run the test for the declare directive's link clause
  test_result = test_declare_link_clause()
  
  IF (test_result) THEN
    PRINT *, "Test for 'declare directive link clause' passed."
    CALL EXIT(0)
  ELSE
    PRINT *, "Test for 'declare directive link clause' failed."
    CALL EXIT(1)
  END IF
END PROGRAM main